gio-subclass/src/auto/action_group.rs

```diff
@@ -0,0 +1,356 @@
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap = from_glib_borrow(gptr);
+    match imp.get_action_state_type(&wrap, &String::from_glib_none(action_name)){ Some(t)  => t/*Not checked*/.to_glib_none().0, None => ptr::null()}
+}
+
+unsafe extern "C" fn actiongroup_has_action<T: ObjectType>
+(gptr: *mut gio_ffi::GActionGroup, action_name: *const c_char) -> gboolean
+{
+    callback_guard!();
+    floating_reference_guard!(gptr);
+    let klass = &**(gptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(gio_ffi::g_action_group_get_type())
+                                     as *const ActionGroupStatic<T>;
+    let instance = &*(gptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap = from_glib_borrow(gptr);
+    imp.has_action(&wrap, &String::from_glib_none(action_name)).to_glib()
+}
+unsafe extern "C" fn actiongroup_list_actions<T: ObjectType>
+(gptr: *mut gio_ffi::GActionGroup) -> *mut *mut c_char
+{
+    callback_guard!();
+    floating_reference_guard!(gptr);
+    let klass = &**(gptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(gio_ffi::g_action_group_get_type())
+                                     as *const ActionGroupStatic<T>;
+    let instance = &*(gptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap = from_glib_borrow(gptr);
+    imp.list_actions(&wrap).to_glib_full()
+}
+unsafe extern "C" fn actiongroup_query_action<T: ObjectType>(
+    gptr: *mut gio_ffi::GActionGroup,
+    action_name: *const c_char,
+    enabled: *mut gboolean,
+    parameter_type: *mut *const glib_ffi::GVariantType,
+    state_type: *mut *const glib_ffi::GVariantType,
+    state_hint: *mut *mut glib_ffi::GVariant,
+    state: *mut *mut glib_ffi::GVariant,
+) -> gboolean {
+    callback_guard!();
+    floating_reference_guard!(gptr);
+    let klass = &**(gptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(gio_ffi::g_action_group_get_type())
+        as *const ActionGroupStatic<T>;
+    let instance = &*(gptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap = from_glib_borrow(gptr);
+    match imp.query_action(&wrap, &String::from_glib_none(action_name)){
+        Some((o_enabled, o_parameter_type, o_state_type, o_state_hint, o_state)) => {
+            ptr::write(enabled, o_enabled.to_glib());
+            ptr::write(parameter_type, o_parameter_type.to_glib_full());
+            ptr::write(state_type, o_state_type.to_glib_full());
+            ptr::write(state_hint, o_state_hint.to_glib_full());
+            ptr::write(state, o_state.to_glib_full());
```

Comment:
The out parameters can all be `NULL` if the caller does not care about a value or multiple. And all pointer-based ones (parameter_type, state_type, state_hint, state) can also be `NULL` themselves (i.e. the Rust side can also return `None` for them)

Reply:
I've regenerated with the current state of the gir. However, shouldn't the return type here be:   

```
Option<(bool, Option<glib::VariantType>, Option<glib::VariantType>, Option<glib::Variant>, Option<glib::Variant>)>
```   

instead of 

```
Option<(bool, glib::VariantType, glib::VariantType, glib::Variant, glib::Variant)>
```

Reply:
Well, everything is wrong :) Your function must return either nothing (action not found), or all of them. And if all of them are returned, you must only write to those pointers that are not `NULL`.

Check the docs of `g_action_group_query_action()`, your function needs to provide exactly the behaviour documented for it

Reply:
Aha. I didn't read it like that at first. Makes sense to not write to a null pointer, indeed :)
Easy enough to fix.